lib/active_merchant/billing/gateways/fortis.rb

```diff
@@ -0,0 +1,330 @@
+          purchase: '/transactions/cc/sale/{placeholder}',
+          capture: '/transactions/{placeholder}/auth-complete',
+          void: '/transactions/{placeholder}/void',
+          refund: '/transactions/{placeholder}/refund',
+          credit: '/transactions/cc/refund/keyed',
+          store: '/tokens/cc',
+          unstore: '/tokens/{placeholder}'
+        }[action]&.gsub('{placeholder}', placeholder.to_s)
+      end
+
+      def payment_type(payment)
+        payment.is_a?(String) ? 'token' : 'keyed'
+      end
+
+      def auth_purchase_request(money, payment, options = {})
+        post = {}
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_address(post, payment, options)
+        post
+      end
+
+      def add_address(post, payment_method, options)
+        address = address(options)
+        return unless address.present?
+
+        post[:billing_address] = {
+          postal_code: address[:zip],
+          street: address[:address1],
+          city: address[:city],
+          state: address[:state],
+          phone: address[:phone],
+          country: lookup_country_code(address[:country])
+        }.compact
+      end
+
+      def address(options)
+        options[:billing_address] || options[:address] || {}
+      end
+
+      def lookup_country_code(country_field)
+        return unless country_field.present?
+
+        country_code = Country.find(country_field)
+        country_code&.code(:alpha3)&.value
+      end
+
+      def add_invoice(post, money, options)
+        post[:transaction_amount] = amount(money)
+        post[:order_number] = options[:order_id]
+        post[:notification_email_address] = options[:email]
+      end
+
+      def add_payment(post, payment, include_cvv = true)
+        case payment
+        when CreditCard
+          post[:account_number] = payment.number
+          post[:exp_date] = expdate(payment)
+          post[:cvv] = payment.verification_value if include_cvv
+          post[:account_holder_name] = payment.name
+        when String
+          post[:token_id] = payment
+        end
+      end
+
+      def parse(body)
+        JSON.parse(body).with_indifferent_access
+      rescue JSON::ParserError, TypeError => e
+        {
+          errors: body,
+          status: 'Unable to parse JSON response',
+          message: e.message
+        }.with_indifferent_access
+      end
+
+      def request_headers
+        CaseSensitiveHeaders.new.reverse_merge!({
+          'Accept' => 'application/json',
+          'Content-Type' => 'application/json',
+          'user-id' => @options[:user_id],
+          'user-api-key' => @options[:user_api_key],
+          'developer-id' => @options[:developer_id]
+        })
+      end
+
+      def add_location_id(post)
+        post[:location_id] = @options[:location_id]
+      end
+
+      def commit(action, post, method = :post, options = {})
+        add_location_id(post) if post.present?
```

Comment:
what is the advantage of sending the `location_id` in the post ❓

Reply:
Is an optional normally that is usefull to identify the request source (Fortis customer normally are shops with card machines or terminals) and for the store transaction is a must.